cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/AbstractStateChangeProgramController.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.program;
+
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.app.runtime.ProgramStateWriter;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.internal.app.runtime.AbstractListener;
+import co.cask.cdap.internal.app.runtime.AbstractProgramController;
+import co.cask.cdap.proto.BasicThrowable;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.id.ProgramId;
+import com.google.common.util.concurrent.Service;
+import org.apache.twill.api.RunId;
+import org.apache.twill.common.Threads;
+import org.apache.twill.internal.ServiceListenerAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Abstract implementation of {@link AbstractProgramController} that responds to state transitions and persists all
+ * state changes.
+ */
+public abstract class AbstractStateChangeProgramController extends AbstractProgramController {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractStateChangeProgramController.class);
+
+  public AbstractStateChangeProgramController(Service service, final ProgramId programId, final RunId runId,
+                                              final ProgramStateWriter programStateWriter,
+                                              @Nullable String componentName) {
+    super(programId, runId, componentName);
+
+    // Add listener to the service for Spark and MapReduce programs
+    if (programId.getType() == ProgramType.MAPREDUCE || programId.getType() == ProgramType.SPARK) {
```

Comment:
`Workers` and `Services` also enter this constructor since their controllers subclass from `ProgramControllerServiceAdapter`. The listeners should not be attached to those `services` because there can be multiple instances. Instead, the listeners are attached to the controller that manages all of those instances. This is done in `InMemoryProgramController`